### ISSUE TITLE
Add "friendly Cython wrappers" for BLAS & LAPACK

### ISSUE THREAD

**Issue by pv** [scipy/scipy#4516]
cf. discussion started in gh-4021

Some prior work exists in 
https://github.com/pv/scipy-work/blob/jason-refactor/scipy/linalg/flapack.pyx.in

**Comment by sturlamolden** [comment]
The "prior art" is wrapping for Python, so requiring the GIL is ok. But it should have been released when calling BLAS/LAPACK.

If we wrap for Cython, we should not require the GIL.

**Comment by jakirkham** [comment]
Isn't this already done?

**Comment by sturlamolden** [comment]
I think this can be closed now.

**Comment by insertinterestingnamehere** [comment]
I'm not sure this should be considered done, though I'm not actively working on it at the moment. The idea was to create memoryview-based and fused type wrappers that let call e.g. `gemm` with memoryviews  rather than `dgemm` with raw pointers. In theory, wrappers using fused types shouldn't be terribly hard to add. Making memoryview-based wrappers would require a lot more work since the wrapping would have to be done on a function-by-function basis.

**Comment by jakirkham** [comment]
Is there a plan for the Cython wrappers to allow for optional arguments? If this is already the case, I may need some pointers to figure out how to use them.

For instance in Python, I could run something like this...

    >>> import numpy as np
    >>> from scipy.linalg import blas
    >>> a = np.ones((100, 101))
    >>> blas.dsyrk(1.0, a)

However, when I try to do something similar in Cython it seems to want 8 more arguments filled out that normally have defaults/are not required.

**Comment by insertinterestingnamehere** [comment]
The current interface is a fairly complete set of raw wrappers around the low-level fortran interface. The arguments are the same as the fortran routines, except that a pointer to each argument is passed rather than the actual value. These are, for the most part, automatically generated from the fortran source files. Unfortunately there's not any clear way to collect the information regarding which arguments should be turned into keyword arguments, which arguments can be passed implicitly as shapes of memoryviews, etc. As far as I'm aware, the only way to do that correctly is by-hand on a routine-by-routine basis. Code generation could still simplify a lot of this, but it can't get around it entirely. I'm not working on this right now, and, to my knowledge, no one else is either.

A much simpler approach for extending these right now might be to modify some of the wrapper generating code to make it so that fused type wrappers around the raw pointer interfaces are created for all of the obvious cases. That would at least make it a little easier for users to write type-generic code. It shouldn't be too hard either, though there are bound to be plenty of corner cases.

**Comment by jakirkham** [comment]
Not sure if this is the right place to ask this, but does code using the Cython BLAS API get linked to the BLAS used by SciPy? Like would this show up if one ran `ldd` on the `*.so` file.

**Comment by rgommers** [comment]
Yes I think that answer is correct. It gets pointers to BLAS functions for the BLAS that SciPy links with, but doesn't need to know/have anything more than that, so should not show up with `ldd`

**Comment by jakirkham** [comment]
Returning to Ian's vision for the BLAS and LAPACK wrappers, agree this would be very nice to see. Though it is hard to automate generating bindings for all of them. Was wondering if we could implement a reasonable subset for BLAS in particular. As it's API has been stable for a long time, this hopefully shouldn't be too bad to hard code solutions for. Any thoughts?

**Comment by insertinterestingnamehere** [comment]
Now that you mention it, maybe a good starting point would be to see if we can write some sort of abbreviated signature for each of the BLAS routines and then generate code to wrap them for memoryviews based off of that. The smaller and more stable interface would provide a nice proving ground for generation techniques that could handle things from LAPACK as well. Starting with BLAS would also be a good approach for generating fused type wrappers.

**Comment by ilayn** [comment]
One detail that is maybe worth mentioning is that Level 1 routine Python wrappers are utilizing a bit nonstandard methods such as offsets. If that difference is not an issue we can try to come up with a signature generator.

**Comment by jakirkham** [comment]
IIUC doing `&a[0]` on a `memoryview` in Cython has already included the offset. So shouldn't be an issue.

One similar thought that occurred to me earlier is the level 2 and level 3 arrays assume a contiguous array. We can admittedly handle that case without too much issue if we enforce striding. Though one thing that might be nice is to detect C or F order and forward this info on to BLAS. Better yet if we can handle cases where the array is not contiguous (e.g. fallback to a slower alternative like a `for`-loop with `dot` or similar).

Actually BLAS is a pretty good test case for this as the function signatures are pretty stereotyped.
